azurerm/internal/services/storage/resource_arm_storage_management_policy.go

```diff
@@ -375,14 +377,20 @@ func flattenStorageManagementPolicyRules(armRules *[]storage.ManagementPolicyRul
 					if armActionBaseBlob.TierToCool != nil && armActionBaseBlob.TierToCool.DaysAfterModificationGreaterThan != nil {
 						intTemp := int(*armActionBaseBlob.TierToCool.DaysAfterModificationGreaterThan)
 						baseBlob["tier_to_cool_after_days_since_modification_greater_than"] = intTemp
+					} else {
+						baseBlob["tier_to_cool_after_days_since_modification_greater_than"] = -1
```

Comment:
No. I already tested. It wouldn't break updates. Could you give me a example?
![image](https://user-images.githubusercontent.com/19754191/76310024-f6386280-6308-11ea-893a-59be947a1d6b.png)

Here there are two issues. 
First issue, the property value in nested block would be set as "0" for TypeInt when tf config doesn't specify this property. But we don't want this is set as 0. We just want this is set as nil. So I have to use this way "d.GetOkExists(fmt.Sprintf("rule.%d.actions.0.base_blob.0.tier_to_archive_after_days_since_modification_greater_than", idx));" to fix this issue.

Second issue, d.set() would set property value in nested block as "0" for TypeInt when service returns nil. So I have to use this way "baseBlob["tier_to_cool_after_days_since_modification_greater_than"] = -1" to fix this issue.

Reply:
The above example doesn't exercise your new code where the values are "-1" in the state (i.e. unset by config).  If you extend your testing to do something like `TestAccAzureRMStorageManagementPolicy_withOneActionWithUpdate` and update the single value set in your new test, I believe you should see what I'm referring to. i.e. The update will read `-1` from the state for the values not explicitly configured which will fail validation and you'll get something like: 

`ManagementPolicy rule rule1 is invalid. Invalid value for parameter : daysAfterModificationGreaterThan`

To avoid this, we should not `d.Set` any value if it's not supplied in the config or read back from the API, meaning the approach for traversing the rules with an index value in `flatten` as you have done for `expand` is necessary. Does that make sense?

Reply:
Seems the update won't read -1 from the state for the values not explicitly configured. The update will read from tf config. Below is my test steps. please correct me if I am not right.

Step1:
`resource "azurerm_resource_group" "test" {
  name     = "storage-mgmt-policy-rg-neil"
  location = "westus"
}

resource "azurerm_storage_account" "test" {
  name                = "storageaccountneil"
  resource_group_name = "${azurerm_resource_group.test.name}"

  location                 = "${azurerm_resource_group.test.location}"
  account_tier             = "Standard"
  account_replication_type = "LRS"
  account_kind             = "BlobStorage"
}

resource "azurerm_storage_management_policy" "test" {
  storage_account_id = "${azurerm_storage_account.test.id}"

  rule {
    name = "RemoveTestContainers"
    enabled = true
    filters {
      blob_types = ["blockBlob"]
      prefix_match = [
        "i-test",
        "p-test",
        "r-test"
      ]
    }
    actions {
      base_blob {
        delete_after_days_since_modification_greater_than = 3
      }
    }
  }
}`

step2:
`resource "azurerm_resource_group" "test" {
  name     = "storage-mgmt-policy-rg-neil"
  location = "westus"
}

resource "azurerm_storage_account" "test" {
  name                = "storageaccountneil"
  resource_group_name = "${azurerm_resource_group.test.name}"

  location                 = "${azurerm_resource_group.test.location}"
  account_tier             = "Standard"
  account_replication_type = "LRS"
  account_kind             = "BlobStorage"
}

resource "azurerm_storage_management_policy" "test" {
  storage_account_id = "${azurerm_storage_account.test.id}"

  rule {
    name = "RemoveTestContainers"
    enabled = true
    filters {
      blob_types = ["blockBlob"]
      prefix_match = [
        "i-test",
        "p-test",
        "r-test"
      ]
    }
    actions {
      base_blob {
        delete_after_days_since_modification_greater_than = 5
      }
    }
  }
}`

Reply:
Yes. I repro the issue what you mentioned. But seems that I cannot remove d.Set for the properties since the properties would be set as "null" when the properties aren't specified in tfconfig. So I've updated code with other solution. Please help to have a look. If the solution is not better, could you share you sample code for the solution? Thanks.

tfstate result:
```
"rule": [
              {
                "actions": [
                  {
                    "base_blob": [
                      {
                        "delete_after_days_since_modification_greater_than": null,
                        "tier_to_archive_after_days_since_modification_greater_than": null,
                        "tier_to_cool_after_days_since_modification_greater_than": null
                      }
                    ],
                    "snapshot": [
                      {
                        "delete_after_days_since_creation_greater_than": 30
                      }
                    ]
                  }
                ],
                "enabled": true,
                "filters": [
                  {
                    "blob_types": [
                      "blockBlob"
                    ],
                    "prefix_match": [
                      "container1/prefix1"
                    ]
                  }
                ],
                "name": "rule1"
              }
            ],
```

tfconfig:
```
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-storage-neil"
  location = "westus"
}

resource "azurerm_storage_account" "test" {
  name                = "unlikelyneiltest"
  resource_group_name = "${azurerm_resource_group.test.name}"
  location                 = "${azurerm_resource_group.test.location}"
  account_tier             = "Standard"
  account_replication_type = "LRS"
  account_kind             = "BlobStorage"
}

resource "azurerm_storage_management_policy" "test" {
  storage_account_id = "${azurerm_storage_account.test.id}"
  rule {
    name    = "rule1"
    enabled = true
    filters {
      prefix_match = ["container1/prefix1"]
      blob_types   = ["blockBlob"]
    }
    actions {
      base_blob {
        delete_after_days_since_modification_greater_than = 102
      }
      snapshot {
        delete_after_days_since_creation_greater_than = 30
      }
    }
  }
}
```